Source/Bricks/Image/ImageBrick.swift

```diff
@@ -194,50 +184,62 @@ open class ImageBrickCell: GenericBrickCell, Bricklike, AsynchronousResizableCel
             return
         }
 
-        if let imageURL = dataSource.imageURLForImageBrickCell(self) {
-            guard currentImageURL != imageURL else {
-                if let image = self.imageView.image {
-                    self.resize(image: image)
-                }
-                return
-            }
-            
+        guard let imageURL = dataSource.imageURLForImageBrickCell(self) else {
             imageView.image = nil
-            currentImageURL = imageURL
+            return
+        }
 
+        // If the requested url is the same as the existing, and the image is already set, then just resize
+        if let image = self.imageView.image, self.currentImageURL == imageURL {
+            self.setImage(image)
+            return
+        }
 
-            self.imageDownloader?.downloadImageAndSet(on: self.imageView, with: imageURL, onCompletion: { (image, url) in
-                self.imageLoaded = true
-                self.resize(image: image)
-                
-                if let delegate = self.brick.delegate, let _ = self.imageView.image {
-                    delegate.didSetImage(brickCell: self)
+        self.currentImageURL = imageURL
+        self.imageDownloader?.downloadImage(with: imageURL, onCompletion: { (image: UIImage, url: URL) in
+            // check again that the url we fetched is the same
+            if self.currentImageURL == url {
+                DispatchQueue.main.async {
+                    self.setImage(image)
                 }
-            })
-        } else {
-            imageView.image = nil
-        }
+            }
+        })
     }
-    
+
+    fileprivate func setImage(_ image: UIImage) {
+        assert(Thread.isMainThread)
+        self.imageLoaded = true
+        self.resize(image: image, onCompletion: {
+            self.imageView.image = image
+            self.brick.delegate?.didSetImage(brickCell: self)
+        })
+    }
+
     override open func prepareForReuse() {
         super.prepareForReuse()
         imageView.image = nil
         currentImageURL = nil
         imageLoaded = false
     }
 
-
-    fileprivate func resize(image: UIImage) {
+    fileprivate func resize(image: UIImage, onCompletion: @escaping (() -> Void)) {
+        self.setRatioConstraint(for: image)
```

Comment:
I don't think we want to set the constraint all the time, It can cause it to be broken

Reply:
fixed

Reply:
The heightContstraint could change depending on the dimension of the image as the cell is re-used. This function removes the exising constraint before adding it again, so I think it should be fine.

Reply:
Yes, but its not needed if the size is fixed. Can we just remove the constraint and then only added it when the height is auto

Reply:
updated